d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java

```diff
@@ -439,117 +439,81 @@ public void testClusterInfoProviderUnregisterClusterListener()
-    state.listenToService = false;
-    SimpleLoadBalancer loadBalancer = spy(setupLoadBalancer(state, serviceRegistry, clusterRegistry, uriRegistry));
+    SimpleLoadBalancerState state =
+        spy(new SimpleLoadBalancerState(new SynchronousExecutorService(), uriRegistry, clusterRegistry, serviceRegistry,
+                                        new HashMap<>(), new HashMap<>()));
+    doAnswer(invocation ->
+             {
+               Thread.sleep(100);
+               return null;
+             }).when(state).listenToService(any(), any());
+    SimpleLoadBalancer loadBalancer = spy(new SimpleLoadBalancer(state, 1, TimeUnit.MILLISECONDS, _d2Executor));
     // case1: listenToService timeout, and simpleLoadBalancer not hit the cache value
-    Callback<ServiceProperties> callback = new Callback<ServiceProperties>()
-    {
-      @Override
-      public void onError(Throwable e)
-      {
-        assertTrue(e instanceof ServiceUnavailableException);
-      }
-
-      @Override
-      public void onSuccess(ServiceProperties result)
-      {
-        throw new RuntimeException("onSuccess should not be called");
-      }
-    };
+    FutureCallback<ServiceProperties> callback = spy(new FutureCallback<>());
     loadBalancer.listenToServiceAndCluster(SERVICE_NAME, callback);
-
-    callback = new Callback<ServiceProperties>()
-    {
-      @Override
-      public void onError(Throwable e)
-      {
-        throw new RuntimeException("onError should not be called");
-      }
-
-      @Override
-      public void onSuccess(ServiceProperties result)
-      {
-        assertTrue(result != null);
-        assertTrue(result.getServiceName().equals(SERVICE_NAME));
-        assertTrue(result.getClusterName().equals(CLUSTER1_NAME));
-      }
-    };
+    // Make sure the onError is called with ServiceUnavailableException only once.
+    verify(callback).onError(any(ServiceUnavailableException.class));
+    verify(loadBalancer).handleTimeoutFromGetServiceProperties(eq(SERVICE_NAME), eq(callback));
 
     // case2: listenToService timeout, and simpleLoadBalancer hit the cache value from state
     LoadBalancerStateItem<ServiceProperties> serviceItem = new LoadBalancerStateItem<>(SERVICE_PROPERTIES, 1, 1);
     when(state.getServiceProperties(SERVICE_NAME)).thenReturn(serviceItem);
     loadBalancer.listenToServiceAndCluster(SERVICE_NAME, callback);
+    // Make sure the onSuccess is called with SERVICE_PROPERTIES only once.
+    verify(callback).onSuccess(eq(SERVICE_PROPERTIES));
 
     // case3: listenToService without timeout
     serviceRegistry.put(SERVICE_NAME, SERVICE_PROPERTIES);
-    loadBalancer = setupLoadBalancer(null, serviceRegistry, clusterRegistry, uriRegistry);
+    doNothing().when(state).listenToService(any(), any());
+    loadBalancer = spy(new SimpleLoadBalancer(state, 5, TimeUnit.SECONDS, _d2Executor));
     loadBalancer.listenToServiceAndCluster(SERVICE_NAME, callback);
+    verify(callback).onSuccess(eq(SERVICE_PROPERTIES));
```

Comment:
This seems incorrect. Calling Mockito `verify` once doesn't clear that invocation, unless it's forcibly cleared. If the success callback was indeed called, this would've needed to be `verify(callback, times(2)).onSuccess(eq(SERVICE_PROPERTIES));`. So this means that `doNothing().` has caused the callback to actually not be invoked.

Reply:
Yes. Thanks for reminding. Have updated the code to `initialize` a new `callback` each time.